server/base/src/main/java/org/apache/accumulo/server/util/Admin.java

```diff
@@ -354,12 +354,8 @@ public void run() {
   }
 
   private static void stopServer(final ClientContext context, final boolean tabletServersToo) throws AccumuloException, AccumuloSecurityException {
-    MasterClient.executeVoid(context, new ClientExec<MasterClientService.Client>() {
-      @Override
-      public void execute(MasterClientService.Client client) throws Exception {
-        client.shutdown(Tracer.traceInfo(), context.rpcCreds(), tabletServersToo);
-      }
-    });
+    MasterClient.executeVoidWithConnRetry(context,
```

Comment:
Is this the only place in the code that is timing out now?   I am wondering this method should have a different name, like `executeVoidWithTimeout()`.  If I am looking at this code, its not clear that this call will timeout.

Reply:
because is not a time but the number of retries that it makes before giving up

Reply:
I suggested making this time based instead of count based in another comment.  That is why I was thinking of that name.  A more general name could be `executeVoidWithLimitedRetries()`.  This name implies it will not retry forever, but does not specify how its limited.  It could be limited by count or time.

Reply:
Yes, we could change the name of it. The 0 is for unlimited retries if you pass like 10 it will try 10 times before it bails out.

Reply:
What if we leave this method as it is (before my commit) and create a new one named for example MasterClient.getConnectionForShutdown(context) this will try two times and return null if it is not possible. This would avoid setting and defining timeouts (For example if some network problem is there for a few hours depending on the time would break things.)
Another question would be: What is a reasonable Timout to set ?

Reply:
I am thinking we may want to provide a way to not retry at all.  I don't think retrying twice vs once offers much benefit.  If it fails the first time it will likely fail the second if retrying immediately.

I have been looking at the code and there is a retry loop in the `executeGeneric()` in addition to the retry loop in `getConnectionWithRetry()`.  I am thinking this code should call an `executeVoid()` method that has no retry loops. Maybe something like the following.

```java
  public static void executeVoid(ClientContext context, ClientExec<MasterClientService.Client> exec, boolean retry)
      throws AccumuloException, AccumuloSecurityException {
    if (retry) {
      executeVoid(context, exec);
    } else {
      MasterClientService.Client client = null;

      try {
        // TODO this gets a connection without a timeout
        client = getConnection(context);
        if (client == null) {
          throw new AccumuloException("Failed to connect to master " + context.getInstance().getMasterLocations());
        }
        exec.execute(client);
      } catch (ThriftSecurityException e) {
        throw new AccumuloSecurityException(e.user, e.code, e);
      } catch (AccumuloException e) {
        throw e;
      } catch (RuntimeException e) {
        throw e;
      } catch (Exception e) {
        throw new AccumuloException(e);
      } finally {
        if (client != null)
          close(client);
      }
    }
  }
```

Reply:
What if we have  some networking issue ?  This way we cannot recover...  
IMHO we should have a call only for the admin tasks so that we can fail directly.

Reply:
> What if we have some networking issue ? This way we cannot recover...

Can you give me an example of the situation you are thinking of?

Reply:
From my understanding this gets called when some client tries to execute a operation on accumulo. Is this correct ? So if I have an application  having some networking timeout (let’s say the cable is out) then it will brake the application and the application will need to take care of retry. At the moment the code does already this. That’s my I think it would be better to have a method only for the admin tasks. Like starting stopping creating tables ... 

Correct me if I’m wrong

Reply:
ok I agree.  I was thinking only the admin task would call `executeVoid(..., false)`

Reply:
Should we agreed on a plan ? Probally having a executeVoidAdmin would be the way to go correct ?

Reply:
executeVoidAdmin is ok with me